### PR TITLE
feat: support pid file.

### PR DESCRIPTION
**Pull request by appleboy**
We can restart the app automatically by the following command.

```sh
kill -USR2 $(cat custom/run/app.pid)
```

### CODE REVIEW

**Comment by lunny** [comment]
Why we need a pid file?

**Comment by appleboy** [comment]
You can also use that information (just do a `cat filename.pid`) to kill the process yourself, using `echo filename.pid | xargs kill`. For develop purpose, you just build the binary and execute `kill -USR2 $(cat custom/run/app.pid)` command to restart app.

**Comment by tboerger** [comment]
I never needed that before but LGTM

**Comment by thibaultmeyer** [comment]
What happening if PID file already exist (ie: the app crash) ? app can be restarted without manually delete PID file ? 

To avoid any issue or conflict with System D, supervisor, or other kind of tools. I think PID file could be created **only**  if -p flag or explicit configuration on app.ini are specified

**Comment by appleboy** [comment]
@0xbaadf00d You don't worry about the PID file exists or not. App can be restarted successfully if you remove the pid file. Just try the following command.

```sh
gitea web
rm -rf custom/run/app.pid
kill -USR2 gitea_pid
cat custom/run/app.pid
```

**Comment by Bwko** [comment]
@appleboy What happens when a startup script also sets a PID?

**Comment by appleboy** [comment]
@Bwko What kind of start script do you use?

**Comment by ekozan** [comment]
what about windows ? it's work ?

**Comment by appleboy** [comment]
@ekozan Yes. Also working on Windows platform.

**Comment by appleboy** [comment]
@Bwko 

screen shot 2017-01-09 at 4 53 55 pm

Also working on [start-stop-daemon](http://man7.org/linux/man-pages/man8/start-stop-daemon.8.html) command

File: modules/setting/setting.go [hunk 1]
```diff
@@ -471,6 +472,19 @@ func IsRunUserMatchCurrentUser(runUser string) (string, bool) {
 	return currentUser, runUser == currentUser
 }
 
+func createPIDFile(pidPath string) {
+	currentPid := os.Getpid()
+	os.MkdirAll(filepath.Dir(pidPath), os.ModePerm)
```
Comment (metalmatze) [inline modules/setting/setting.go]: Please check the error returned by `os.MkdirAll()`.

**Review by metalmatze — CHANGES_REQUESTED** [review]

**Comment by Bwko** [comment]
@appleboy Thanks :+1:  When you add metalmatze suggestions I'll LGT-M :smile:

File: modules/setting/setting.go (hunk 1 above)
Comment (appleboy) [reply modules/setting/setting.go]: @metalmatze Done! Please review again.

**Comment by appleboy** [comment]
@Bwko Done!

File: modules/setting/setting.go [hunk 2]
```diff
@@ -471,6 +472,22 @@ func IsRunUserMatchCurrentUser(runUser string) (string, bool) {
 	return currentUser, runUser == currentUser
 }
 
+func createPIDFile(pidPath string) {
+	currentPid := os.Getpid()
+	if err := os.MkdirAll(filepath.Dir(pidPath), os.ModePerm); err != nil {
+		log.Fatal(4, "Can' create PID folder on %s", err)
+	}
+
+	file, err := os.Create(pidPath)
```
Comment (Bwko) [inline modules/setting/setting.go]: Quoting bkcsoft on this one:

> Overall, change the format to this instead whenever possible. Keeps errors(variables) bound to their context(block).

`if err := os.Create(...); err != nil {
  [...]
}`

File: modules/setting/setting.go [hunk 3]
```diff
@@ -471,6 +472,22 @@ func IsRunUserMatchCurrentUser(runUser string) (string, bool) {
 	return currentUser, runUser == currentUser
 }
 
+func createPIDFile(pidPath string) {
+	currentPid := os.Getpid()
+	if err := os.MkdirAll(filepath.Dir(pidPath), os.ModePerm); err != nil {
+		log.Fatal(4, "Can' create PID folder on %s", err)
```
Comment (Bwko) [inline modules/setting/setting.go]: Some nitpicking: missing a `t` here

File: modules/setting/setting.go (hunk 2 above)
Comment (appleboy) [reply modules/setting/setting.go]: @Bwko We need `file` variable.

Comment (Bwko) [reply modules/setting/setting.go]: Sorry my bad :sweat_smile:

**Comment by Bwko** [comment]
LGTM

**Comment by lunny** [comment]
need @metalmatze approve

**Comment by appleboy** [comment]
@lunny @Bwko @metalmatze Thanks

**Pull request merged**